File: Sources/ChatUnpackCore/Domain/ScrollPosition.swift
import Foundation

public struct ScrollPosition: Sendable, Equatable {
  public let value: CGFloat
  public let minimum: CGFloat
  public let maximum: CGFloat

  public init(value: CGFloat, minimum: CGFloat, maximum: CGFloat) {
    self.value = value
    self.minimum = minimum
    self.maximum = maximum
  }

  public var normalized: CGFloat {
    guard isUsable else { return 0 }
    return min(1, max(0, (value - minimum) / (maximum - minimum)))
  }

  public var isUsable: Bool {
    value.isFinite
      && minimum.isFinite
      && maximum.isFinite
      && maximum - minimum > 0.0001
      && value >= minimum - 0.01
      && value <= maximum + 0.01
  }

  public var isAtBottom: Bool {
    isUsable && normalized >= 0.99
  }
}

File: windows/src/ChatUnpack.Core/Domain/ScrollPosition.cs
namespace ChatUnpack.Core.Domain;

public readonly record struct ScrollPosition
{
  public ScrollPosition(double value, double minimum, double maximum)
  {
    Value = value;
    Minimum = minimum;
    Maximum = maximum;
  }

  public double Value { get; }
  public double Minimum { get; }
  public double Maximum { get; }

  public double Normalized
  {
    get
    {
      if (!IsUsable)
      {
        return 0;
      }

      return Math.Min(1, Math.Max(0, (Value - Minimum) / (Maximum - Minimum)));
    }
  }

  public bool IsUsable
  {
    get
    {
      return double.IsFinite(Value)
        && double.IsFinite(Minimum)
        && double.IsFinite(Maximum)
        && Maximum - Minimum > 0.0001
        && Value >= Minimum - 0.01
        && Value <= Maximum + 0.01;
    }
  }

  public bool IsAtBottom => IsUsable && Normalized >= 0.99;
}
